easybuild/tools/toolchain/compiler.py

```diff
@@ -291,14 +291,17 @@ def _set_optimal_architecture(self, default_optarch=None):
 
             # optarch has been validated as complex string with multiple compilers and converted to a dictionary
             elif isinstance(optarch, dict):
-                current_compiler = getattr(self, 'COMPILER_FAMILY', None)
-                if current_compiler in optarch:
-                    if optarch[current_compiler] == OPTARCH_GENERIC:
-                        use_generic = True
-                    else:
+                # first try module names, than the family in optarch
+                current_compiler_names = (getattr(self, 'COMPILER_MODULE_NAME', []) +
+                                          [getattr(self, 'COMPILER_FAMILY', None)])
+                for current_compiler in current_compiler_names:
+                    if current_compiler in optarch:
                         optarch = optarch[current_compiler]
+                        break
+                if optarch == OPTARCH_GENERIC:
+                    use_generic = True
                 # no option for this compiler
-                else:
+                if isinstance(optarch, dict):
```

Comment:
I find it odd the recheck for `optarch` being a `dict`. It makes sense, but I'd clarify it in the comment above. Something like:

`# optarch hasn't changed, it is still a dict, which means that there is no option for this compiler`